### PR TITLE
Add mcu::HWC service

### PR DESCRIPTION
**Pull request by Pirater12**

### CODE REVIEW

File: libctru/include/3ds.h [hunk 1]
```diff
@@ -27,7 +27,7 @@ extern "C" {
 #include <3ds/allocator/linear.h>
 #include <3ds/allocator/mappable.h>
 #include <3ds/allocator/vram.h>
-
+
```
Comment (fincs) [inline libctru/include/3ds.h]: Spurious whitespace change

File: libctru/include/3ds/services/mcuHWC.h [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+/// Exits mcuHWC.
+void mcuHWCExit(void);
+
+/**
+ * @brief Reads data from a MCU Register
+ * @param reg Register number. See https://www.3dbrew.org/wiki/I2C_Registers#Device_3 for more info
+ * @param size Size of data to be read
+ * @param data Pointer to write the data to.
+ */
+Result mcuReadRegister(u8 reg, u32 size, u8* data);
```
Comment (fincs) [inline libctru/include/3ds/services/mcuHWC.h]: void*

File: libctru/include/3ds/services/mcuHWC.h [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+/**
+ * @brief Writes data to a MCU Register
+ * @param reg Register number. See https://www.3dbrew.org/wiki/I2C_Registers#Device_3 for more info
+ * @param size Size of data to be written
+ * @param data Pointer to write the data from.
+ */
+Result mcuWriteRegister(u8 reg, u32 size, u8* data);
```
Comment (fincs) [inline libctru/include/3ds/services/mcuHWC.h]: const void*

File: libctru/include/3ds/services/mcuHWC.h [hunk 4]
```diff
@@ -0,0 +1,46 @@
+/**
+ * @file mcuHWC.h
+ * @brief mcuHWC service.
+ */
+#pragma once
+
+/// Initializes mcuHWC.
+Result  mcuHWCInit(void);
```
Comment (fincs) [inline libctru/include/3ds/services/mcuHWC.h]: Double space -> single space

File: libctru/source/services/mcuHWC.c [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+
+Result mcuReadRegister(u8 reg, u32 size, u8* data)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x1,2,2); // 0x10082
+	cmdbuf[1] = reg;
+	cmdbuf[2] = size;
+	cmdbuf[3] = size << 4 | 0xC;
```
Comment (fincs) [inline libctru/source/services/mcuHWC.c]: Use the corresponding IPC macro to build this translation header

File: libctru/source/services/mcuHWC.c [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+
+Result mcuWriteRegister(u8 reg, u32 size, u8* data)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x2,2,2); // 0x20082
+	cmdbuf[1] = reg;
+	cmdbuf[2] = size;
+	cmdbuf[3] = size << 4 | 0xA;
```
Comment (fincs) [inline libctru/source/services/mcuHWC.c]: Use the corresponding IPC macro to build this translation header

File: libctru/include/3ds.h (hunk 1 above)
Comment (WinterMute) [inline libctru/include/3ds.h]: can you strip trailing whitespace in your changes, please?

File: libctru/source/services/mcuHWC.c [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+	
+	return (Result)cmdbuf[1];
+}
+
+Result mcuWriteRegister(u8 reg, u32 size, u8* data)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace here

File: libctru/source/services/mcuHWC.c [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x1,2,2); // 0x10082
+	cmdbuf[1] = reg;
+	cmdbuf[2] = size;
+	cmdbuf[3] = size << 4 | 0xC;
+	cmdbuf[4] = (u32)data;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x1,2,2); // 0x10082
+	cmdbuf[1] = reg;
+	cmdbuf[2] = size;
+	cmdbuf[3] = size << 4 | 0xC;
+	cmdbuf[4] = (u32)data;
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x2,2,2); // 0x20082
+	cmdbuf[1] = reg;
+	cmdbuf[2] = size;
+	cmdbuf[3] = size << 4 | 0xA;
+	cmdbuf[4] = (u32)data;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x2,2,2); // 0x20082
+	cmdbuf[1] = reg;
+	cmdbuf[2] = size;
+	cmdbuf[3] = size << 4 | 0xA;
+	cmdbuf[4] = (u32)data;
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0xB,0,0); // 0xB0000
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+	
+	*level = cmdbuf[2];
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+Result mcuGetSoundSliderLevel(u8 *level)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0xB,0,0); // 0xB0000
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+	return (Result)cmdbuf[1];
+}
+
+Result mcuGetSoundSliderLevel(u8 *level)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0xB,0,0); // 0xB0000
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+	*level = cmdbuf[2]; 
+	
+	return (Result)cmdbuf[1];
+}
+
+Result mcuGetSoundSliderLevel(u8 *level)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+{
+	if (AtomicDecrement(&mcuHWCRefCount)) return;
+	svcCloseHandle(mcuHWCHandle);
+}
+
+Result mcuReadRegister(u8 reg, u32 size, u8* data)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+	
+	return (Result)cmdbuf[1];
+}
+
+Result mcuGetBatteryVoltage(u8 *voltage)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+	return (Result)cmdbuf[1];
+}
+
+Result mcuGetBatteryVoltage(u8 *voltage)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x4,0,0); // 0x40000
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+Result mcuGetBatteryVoltage(u8 *voltage)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x4,0,0); // 0x40000
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x4,0,0); // 0x40000
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+	
+	*voltage = cmdbuf[2]; 
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+	*voltage = cmdbuf[2]; 
+	
+	return (Result)cmdbuf[1];
+}
+
+Result mcuGetBatteryLevel(u8 *level)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 22]
```diff
@@ -0,0 +1,93 @@
...
+	return (Result)cmdbuf[1];
+}
+
+Result mcuGetBatteryLevel(u8 *level)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x5,0,0); // 0x50000
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 23]
```diff
@@ -0,0 +1,93 @@
...
+Result mcuGetBatteryLevel(u8 *level)
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x5,0,0); // 0x50000
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

File: libctru/source/services/mcuHWC.c [hunk 24]
```diff
@@ -0,0 +1,93 @@
...
+{
+	Result ret = 0;
+	u32 *cmdbuf = getThreadCommandBuffer();
+	
+	cmdbuf[0] = IPC_MakeHeader(0x5,0,0); // 0x50000
+	
+	if(R_FAILED(ret = svcSendSyncRequest(mcuHWCHandle)))return ret;
+	
+	*level = cmdbuf[2]; 
+
```
Comment (WinterMute) [inline libctru/source/services/mcuHWC.c]: trailing whitespace

**Comment by joel16** [comment]
Okay well clearly you got this from 3DBrew, thanks to sono.

And also, for consistency purposes shouldn't the functions names be like MCU_doSomething()?

**Comment by Pirater12** [comment]
@fincs do functions also need to begin with mcuHwc ?

**Comment by fincs** [comment]
Sorry for taking so long in reviewing this.
Functions should be prefixed with `mcuHwc` (i.e. NOT `mcuHWC` or `mcu`).
Files should be named `mcuhwc.h/c` (i.e. NOT `mcuHwc.h/c`)

**Comment by joel16** [comment]
@fincs I would still like to know why this isn't consistent with the other service api functions?

Example GSPLCD_GetVendors, ACU_GetWifiStatus etc.

Shouldn't they have MCU capitalized like I had them here: https://github.com/joel16/3DSident/blob/master/source/services/mcu.c

**Comment by fincs** [comment]
Tbh it's not even consistent within libctru, see [boss.h](https://github.com/smealum/ctrulib/blob/master/libctru/include/3ds/services/boss.h).

**Comment by fincs** [comment]
We're in the middle of a release process right now, trying to wrap things up.

**Comment by joel16** [comment]
Hate to continue an off topic discussion here, but I wanted to ask if you'd consider waiting for a few more AM functions that weren't included: https://gist.github.com/joel16/5b432610add4e9b7dc04216d4e31df92

**Pull request merged**